### ISSUE TITLE
ReduceSum operation should support multiple axes

### ISSUE THREAD

**Issue by vuzelac-cadence** [pytorch/glow#2343]
Running ResNet v1 50 model obtained from TF2ONNX and it fails with:
      `"Only supporting single reduction for now."`
This originates from `ReduceSum` operation processing in` CommonOperatorLoader.h` from `Importer`

I assume the support for multiple axis is to be added to `BatchedReduceAddNode`, however, I am curious why a simpler implementation has not being done to at least support ReduceSum w/ multiple axes until there is a better implementation:
For example, repeating current code in a for loop for each dimension, to emit serialized `BatchedReduceAddNode`  nodes (or serialized `BatchedReduceAddNode`, `Splat`, `Divide`, for `ReduceMean)` seems trivial yet enables more use cases.

**Comment by jfix71** [comment]
Hi @vuzelac-cadence, we just haven't hit a use case for it and so we left it unimplemented. As you said it should be relatively easy to support. There are a few options here. One could directly implement it via multiple nodes in `Graph.cpp`. Another option is to modify the current `BatchedAddReduceNode`, and then either lower it to multiple `BatchedAddReduceNodes`, or modify the kernels in each backend that supports it.

**Comment by vuzelac-cadence** [comment]
First want to mention that the issue here is with ResNets v1 50/152, coming from TensorFlow and TF2ONNX converter - they do have ReduceMean w/ multiple dimensions. Some ONNX converters will blindly replace ReduceMean with AveragePool which is a bit of a patch, but still it can be done in Glow as well (see option 2 below).

I started with a simple serialization of the current ReduceMean implementation, for each dimension, but that doesn't seem like a good option; For example, when AveragePool is a good candidate to replace ReduceMean, it would be rather hard to evaluate add+splat+div+add+splat+div in the backend, and replace it all with AveragePool.

We propose two solutions: 
1.) Introduce ReduceMean node, that will be lowered in a lower layer, not in the Importer.

2.) Replace ReduceMean with AveragePool node right away in the Importer (if it can be done - keepdims=1 and 2 last axes, W&H, are reduced).  A backend might not be able to handle large size kernels but that's not any different from what backend has to do if it doesn't support an existing  AveragePool node. 

Option 2) is much faster to implement, but is a bit of a patch. It does affects backends that do support ReduceMean (w/ multiple axes) which would likely want to simple receive a ReduceMean node, as described in 1).  This option might be a good choice if focusing on just enabling ResNets coming from TF.

**Comment by jfix71** [comment]
@vuzelac-cadence Thanks for the analysis. I agree that option 2 is more of a patch. I'd prefer option 1, especially now that our quantization profiling will profile the lowered components of nodes that are lowered https://github.com/pytorch/glow/pull/2311. I think many of our current nodes which are implemented directly in `Graph.cpp`/without a node in `NodeGen.cpp` were done mostly to avoid this (now solved) issue.

It shouldn't be too hard to implement option 1. You basically move the logic from `Function::createBatchedReduceMean()` into a case in `Lower.cpp`, and add the new node to `NodeGen.cpp`, and add verification logic to `Nodes.cpp`, and that's basically it.

Also, if you end up implementing the `ReduceMean -> AvgPool` graph optimization you mentioned, it would be great to have it upstreamed! 😃

**Comment by vuzelac-cadence** [comment]
I have opened a PR (#2465 ) that fails in a weird way in an OpenCL test (newly added one). We don't have OpenCL setup (yet)  to experiment but anyway, it would be great if somebody can take a look into my short description there, as what OpenCL expects definitively looks wrong.
